### PR TITLE
:tada: New Source: Parcel Perform

### PR DESCRIPTION
**Pull request by apostoltego**
## What
Source for Parcel Perform 
Related to this [issue](https://github.com/airbytehq/hacktoberfest-2021/issues/61) 
## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by apostoltego** [comment]
Hey @avaidyanatha , @marcosmarxm there's some concerns with getting this connector live:

- I believe for the hackathon you needed sandbox account etc. I’ve spoken to parcelperform and they’ve generated one under our main account however this has an expiration date. Secondary to this is that you need real parcels (shipments) to track through this tool so even with a secondary account this becomes a bit of a problem in both generating data and somehow have a public set of creds (the shipments would be holding some recipients address etc)
- Tried getting the caching mechanism to work but could never actually figure out why when  use_cache = True the parent object gets called. In this case it was creating without state passed through which in turn triggers a sync from config’s start date.

**Comment by apostoltego** [comment]
Screenshot 2021-11-12 at 19 46 11
Integration tests

**Comment by marcosmarxm** [comment]
Hello @apostoltego ! Sorry to not finished your contribution before the date stipulated in the contest. All contributions made before 15-November are eligible to receive the award. We're trying to review your contribution as soon as possible.

**Comment by apostoltego** [comment]
@marcosmarxm  Don't worry about it - still keen to get this one through. Especially with some guidance on the concerns raised in the first comment.

**Comment by harshithmullapudi** [comment]
Hey @apostoltego can you move this PR out of Draft state when you are done with this work.

**Comment by apostoltego** [comment]
hey @harshithmullapudi I've left in draft due to the concerns raised above but on my end there's no immediate next steps. Any chance you can review and suggest whether you'd like to implement caching or just proceed as is?

File: airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            response.raise_for_status()
+            response_json = response.json()
+            return response_json["access_token"], response_json["expires_in"]
+        except Exception as e:
+            raise Exception(f"Error while refreshing access token: {e}") from e
+
+
+class Shipments(HttpStream, ABC):
+
+    url_base = "https://api.parcelperform.com"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py:57]: Add the url_base in ParcelSource

File: airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        self, response: requests.Response
+    ) -> Optional[Mapping[str, Any]]:
+        """
+        :param response: the most recent response from the API
+        :return If there is another page in the result, a mapping (e.g: dict) containing information needed to query the next page in the response.
+                If there are no more pages in the result, return None.
+        """
+        if response.status_code == 200:
+            return response.json()["next_page"]
+        elif response.status_code == 404 and response.json()["api_response"] == "4041":
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py:102]: the code 4041 is correct? can you add a documentation explaining this?

File: airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        :return Tuple[bool, any]: (True, None) if the input config can be used to connect to the API successfully, (False, error) otherwise.
+        """
+        auth = ParcelPerformAuthenticator(
+            "https://api.parcelperform.com/auth/oauth/token/",
+            config["client_id"],
+            config["client_secret"],
+            "access_token",
+        )
+        if auth.get_access_token():
+            return True, None
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py:216]: See other connectors `check_connection` function example. Most cases you need a try/except

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
See comments @apostoltego thanks for the contribution. I requested our team to get one sandbox account to test your contribution.

**Comment by marcosmarxm** [comment]
@apostoltego please sign the CLA

**Comment by apostoltego** [comment]
@marcosmarxm  Thanks for the feedback. Your comments should be addressed. Regarding the CLA not sure why it says pending - when I click through it shows that I've signed it.

**Comment by alafanechere** [comment]
Hi @apostoltego, could you set your PR as ready for review if you think you've made the requested changes. Moreover the CLA signature seems to remain pending. It looks like CLA assistant cannot match your username with a github user, could you please troubleshoot this with this [documentation](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user)?

**Comment by apostoltego** [comment]
Done, thanks for the help @alafanechere

**Comment by karinakuz** [comment]
Free Trial account is created (expiration date 2021-12-12), OAuth token is received, credentials are added to the LastPass with the name "Parcel Perform". APIs are tested for two streams: 
- Shipments List https://api.parcelperform.com/v5/shipment/list/?created_date_from=2021-11-01T10:30:00&created_date_to=2021-11-30T10:30:00&carrier_reference=usps
- Shipment Details https://api.parcelperform.com/v5/shipment/details/?shipment_id=shipment001&shipment_uuid=a0fe1b96-3932-4d91-b9b4-a7073c72236b

File: airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/schemas/shipments_details.json [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+              "place": {
+                "type": ["null", "string"]
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/schemas/shipments_details.json:371]: can you add a new line ?

File: airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        return {
+            self.cursor_field: max(
+                latest_record.get(self.cursor_field),
+                current_stream_state.get(self.cursor_field, self.config["start_date"]),
+            )
+        }
+
+
+class Shipments(ParcelPerformStream, ABC):
+
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py:96]: can you add link(in comments) to the api documentation for this stream ?

File: airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+                headers=self.request_headers(),
+            )
+            response.raise_for_status()
+            response_json = response.json()
+            return response_json["access_token"], response_json["expires_in"]
+        except Exception as e:
+            raise Exception(f"Error while refreshing access token: {e}") from e
+
+
+class ParcelPerformStream(HttpStream, ABC):
```
Comment (harshithmullapudi) [inline airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py:55]: can you add link(in comments) to the api documentation for this stream ?

**Review by harshithmullapudi — CHANGES_REQUESTED** [review]

**Comment by harshithmullapudi** [comment]
Hey @apostoltego  can you run `./gradlew format` in the root airbyte folder. We have also got credentials so we are good to go. Thanks for your patience.

**Review by alafanechere — CHANGES_REQUESTED** [review]
Hi @apostoltego I want to give you a first code review but I'd like the acceptance test to pass before: 
`./gradlew :airbyte-integrations:connectors:source-parcel-perform:integrationTest` must run without failures.

Here are the failures I encountered:
* Flake checks: You've some unused import
* Unit tests: the test declared in `airbyte-integrations/connectors/source-parcel-perform/unit_tests` are not passing, please make them pass or update them according to your connector logic
* Integration tests: 
  * connection test with `status == "failed"` should run with a valid config in term of schema but with wrong credentials in `integration_tests/invalid_config.json` the date is not valid according to you spec.
  * the `basic_read` is failing with multiple errors of this kind: 
 ```
Failed validating 'type' in schema['properties']['return_address']:
    {'type': ['null', 'string']}

On instance['return_address']:
    {'address_name': 'org-001',
     'city': 'Chiyoda-ku, Tokyo',
     'company': 'Warehouse Manufacturing Company',
     'country': 'Japan',
     'country_code': 'JP',
     'email': 'email@warehouse.com',
     'entity_reference': 's001-org-001',
     'first_name': '',
     'full': '7-30, Marunouchi 2-chome, #01-55, 100-6432, Chiyoda-ku, '
             'Tokyo, Kanto, Japan',
     'last_name': '',
     'line1': '7-30',
     'line2': 'Marunouchi 2-chome',
     'line3': '#01-55',
     'location_type': 'warehouse',
     'phone': '+81959820269',
     'postal_code': '100-6432',
     'region': 'Kanto',
     'salutation': '',
     'state_or_province': '',
     'tax_id': 'TXA51-31'}
```
Please update your configured catalog to handle these nested object type.

**Comment by marcosmarxm** [comment]
@apostoltego  I'll close this due to inactivity. If you want to return this work you can reopen the PR.

**Pull request closed without merging**